text/0000-separate-context-propagation.md

```diff
@@ -0,0 +1,193 @@
+
+**GetCurrent() -> context**  
+To access the context associated with program execution, the Context API provides a function which takes no arguments and returns a Context.
+
+
+## Propagation API
+
+Distributed applications send data to downstream processes via propagators, functions which read and write application context into RPC requests. Each distributed application creates a set of propagators for every type of supported medium - currently HTTP and Binary.
+
+**Inject(context, request)**  
+To send the data for all distributed applications downstream to the next process, the Propagation API provides a function which takes a context and a request, and mutates the request to include the encoded context. The canonical representation of a request is as a map.
+
+**Extract(context, request) -> context**  
+To receive data injected by prior upstream processes, the Propagation API provides a function which takes a context and a request, and returns an updated context.
+
+**RegisterPropagator(type, inject, extract)**  
+In order for the application layer to function correctly, Propagation choices must be syncronized between all processes in the distributed system, and multiple applications must be able to inject and extract their context into the same request. To meet these requirements, the Propagation API provides a function which registers a set of propagators, which will all be executed in order when the future calls to inject and extract are made. A canonical propagator consists of an inject and an extract function.
+
+OpenTelemetry currently contains two types of Propagators:
+
+* **HTTP** - context is written into and read from a map of HTTP headers.
+* **Binary** - context is serialized into and deserialized from a stream of bytes.
+
+# Internal details
+
+![drawing](img/context_propagation_details.png)
+
+## Context details
+OpenTelemetry currently implements three context types of context propagation.
+
+**Span Context -** The serializable portion of a span, which is injected and extracted. The readable attributes are defined to match those found in the  W3C **traceparent** header. 
+
+**Correlation Context -** Correlation Context contains a map of labels and values, to be shared between metrics and traces. This allows observability data to be indexed and dimensionalized in a variety of ways. Note that correlations can quickly add overhead when propagated in-band. But because this data is write-only, it may be possible to optimize how it is transmitted.
+
+**Baggage Context -** Transaction-level application data, meant to be shared with downstream components. This data is readable, and must be propagated in-band. Because of this, Baggage should be used sparingly, to avoid ballooning the size of RPC requests.
+
+Note that when possible, OpenTelemetry APIs calls are given access to the entire context object, and not a specific context type.
```

Comment:
I like the whole context object as it allows more flexibility for the propagators to modify contexts, but there's implementation details that all propagators will have to be aware of:

1. how to get and set specific contexts from the Context object. One way to accomplish this is with an explicit interface or methods attached to context, but as I understand it, this violates the idea that Yuri was evangelizing where the Context object is a primitive and is completely unaware of what values are being loaded into it.

2. SpanContext is a little tricky where the extractor should be setting the parent SpanContext, while the injector should be injecting values from the current SpanContext. This would be great to not require the propagator author to get that right. 

I'm personally voting for an explicit interface on context to retrieve such values. solving #2 is a little trickier. 

More context in this discussion:
https://github.com/open-telemetry/oteps/pull/37
And this PR:
https://github.com/open-telemetry/opentelemetry-python/pull/89

Reply:
@toumorokoshi in Go you can define a private namespace for the values you store in Context ([example in go][1]). Nobody else but your cross-cutting application needs to know how to get/set those values. Therefore, the Context can remain completely agnostic to what it stores and carries.

[1]: https://github.com/opentracing/opentracing-go/blob/master/gocontext.go#L5

Reply:
Ok, I think I understand what you're proposing. It's

1. a property-agnostic context object.
2. a layer of getter and setters to get and set values from that context. 

I'm fine with this approach. We actually had a discussion in the python group and that's the consensus we've reached as well.